bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/handler/VerisureThingHandler.java

```diff
@@ -0,0 +1,196 @@
+        config = getConfigAs(VerisureThingConfiguration.class);
+        // Set status to UNKNOWN and let background task set correct status
+        updateStatus(ThingStatus.UNKNOWN);
+        scheduler.execute(() -> {
+            Bridge bridge = getBridge();
+            if (bridge != null) {
+                this.bridgeStatusChanged(bridge.getStatusInfo());
+            }
+        });
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("dispose on thing: {}", thing);
+        Bridge bridge = getBridge();
+        if (bridge != null) {
+            VerisureBridgeHandler vbh = (VerisureBridgeHandler) bridge.getHandler();
+            if (vbh != null) {
+                session = vbh.getSession();
+                if (session != null) {
+                    session.unregisterDeviceStatusListener(this);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        logger.debug("bridgeStatusChanged bridgeStatusInfo: {}", bridgeStatusInfo);
+        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE) {
+            Bridge bridge = getBridge();
+            if (bridge != null) {
+                VerisureBridgeHandler vbh = (VerisureBridgeHandler) bridge.getHandler();
+                if (vbh != null) {
+                    session = vbh.getSession();
+                    String deviceId = config.getDeviceId();
+                    if (session != null) {
+                        update(session.getVerisureThing(deviceId));
+                        session.registerDeviceStatusListener(this);
+                    }
+                }
+            }
+        }
+        super.bridgeStatusChanged(bridgeStatusInfo);
+    }
+
+    @Override
+    public void onDeviceStateChanged(VerisureThing thing) {
+        logger.trace("onDeviceStateChanged on thing: {}", thing);
+        String deviceId = thing.getDeviceId();
+        // Make sure device id is normalized, i.e. replace all non character/digits with empty string
+        if (config.getDeviceId().equalsIgnoreCase((deviceId.replaceAll("[^a-zA-Z0-9]+", "")))) {
+            update(thing);
+        }
+    }
+
+    public synchronized void update(@Nullable VerisureThing thing) {
+        ChannelUID cuid = new ChannelUID(getThing().getUID(), CHANNEL_INSTALLATION_ID);
+        if (thing != null) {
+            BigDecimal siteId = thing.getSiteId();
+            updateState(cuid, new DecimalType(siteId.longValue()));
+            cuid = new ChannelUID(getThing().getUID(), CHANNEL_INSTALLATION_NAME);
+            updateState(cuid, new StringType(thing.getSiteName()));
+        }
+    }
+
+    @Override
+    public void onDeviceRemoved(@Nullable VerisureThing thing) {
+        logger.trace("onDeviceRemoved on thing: {}", thing);
+    }
+
+    @Override
+    public void onDeviceAdded(@Nullable VerisureThing thing) {
+        logger.trace("onDeviceAdded on thing: {}", thing);
+    }
```

Comment:
These method seem to be meant to be overridden by subclasses. Given that each handler subclass is meant to handle a particular VerisureThing subclass I think it should be enforced through generics. It looks like the subclasses are forced to check the thing instance type for every method. To eliminate this redundancy you will need to start by add a generic parameter to DeviceStatusListener:
```
public interface VerisureThingHandler<T extends VerisureThing>{
    public void onDeviceStateChanged(T thing);
    public void onDeviceRemoved(T thing);
    public void onDeviceAdded(T thing);
    public Class<T> getVerisureThingClass();
}
```
You will see that I added an extra method but you will see why later.
Next, lets add a generic parameter to the VerisureThingHandler definition:
```
public abstract class VerisureThingHandler<T extends VerisureThing>  extends BaseThingHandler implements DeviceStatusListener<T> {
    @Override
    public void onDeviceStateChanged(T thing) {
//logic
    }
    @Override
    public void onDeviceRemoved(T thing) {
        logger.trace("onDeviceRemoved on thing: {}", thing);
    }
    @Override
    public void onDeviceAdded(T thing) {
        logger.trace("onDeviceAdded on thing: {}", thing);
    }
    public synchronized void update(T thing) {
    //logic
    }
}
```
Handler subclasses would then only have to worry about dealing with thing types they are meant to deal with.

Now the only thing left to do would be to add the necessary logic to your session class to make sure that update event only get forwarded to listeners of the correct class type. Now this is where that `getVerisureThingClass()` method from earlier will become necessary. The session will need to call that method for each listener and compare it against the thing class before calling the listener's update methods.

Reply:
I'll do a push of my updated code upto this comment before doing this change since this might take a while fix. :-)

Reply:
@cpmeister My hat off, your proposal is now implemented and running on my production environment. :-)